qp/utils.py

```diff
@@ -0,0 +1,46 @@
+import numpy as np
+import sys
+
+def safelog(arr, threshold=sys.float_info.epsilon):
+    """
+    Takes log of array with zeroes.
+
+    Parameters
+    ----------
+    arr: ndarray
+        Values to be logged
+    threshold: float
+        Small, positive value to replace zeros and negative numbers
+
+    Returns
+    -------
+    logged: ndarray
+        Logged values, with approximation in place of zeros and negative numbers
+    """
+    shape = np.shape(arr)
+    flat = arr.flatten()
+    logged = np.log(np.array([max(a,threshold) for a in flat])).reshape(shape)
+    return logged
+
+def calckl(qn, pn):
+    """
+    Calculates Kullback-Leibler Divergence
+
+    Parameters
+    ----------
+    qn: ndarray
```

Comment:
No, `P` and `Q` need to be any PDF objects. What I would do is enable the statement `D = qp.KL(P, Q)` by having the `KL` function use the `evaluate` method of each PDF object. `evaluate` should call `pdf()` if `truth is not None`, and otherwise call `approximate()` with whatever it has to hand: first try the quantiles, then the histogram (later!). Then, you make a `P` that has a `truth` but no `quantiles` and a `Q` that has `quantiles` but no truth. What do you think?